### PR TITLE
Ensure cleanup happens when enabling --warm or --autoreload with Bokeh 2.4

### PR DESCRIPTION
**Pull request by philippjfr**
Fixes https://github.com/holoviz/panel/issues/2755

### CODE REVIEW

File: panel/tests/command/test_serve.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+            pytest.fail("no matching log line in process output")
+        port = int(m.group(1))
+
+        r = requests.get(f"http://localhost:{port}/{app_name}")
+        assert r.status_code == 200
+        assert "<title>A</title>" in r.content.decode('utf-8')
+
+        py_file.file.write(app2)
+        py_file.file.flush()
+
```
Comment (maximlt) [inline panel/tests/command/test_serve.py:104]: Shouldn't it wait here for a little while?

File: panel/tests/command/test_serve.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    try:
+        yield p
+    except Exception as e:
+        p.terminate()
+        p.wait()
+        print("An error occurred: %s", e)
+        try:
+            out = p.stdout.read().decode()
+            print("\n---- subprocess stdout follows ----\n")
+            print(out)
```
Comment (maximlt) [inline panel/tests/command/test_serve.py:27]: Are these print statements captured by pytest and displayed when running the tests?

Comment (philippjfr) [reply panel/tests/command/test_serve.py:27]: They are if the test fails.

File: panel/tests/command/test_serve.py (hunk 1 above)
Comment (philippjfr) [reply panel/tests/command/test_serve.py:104]: Flushing the write buffer should really be sufficient but who knows what it's doing on some platforms.

**Comment by Hoxbro** [comment]
I'm seeing some (new) problems with this fix.

I have this code 
``` python
import panel as pn
pn.panel("Test").servable()
```
Which I run with `panel serve race_condition.py --autoreload`, but when I reload the page multiple times I get this error:

``` bash
2021-09-20 22:06:31,569 DocumentLifeCycleHandler on_session_destroyed callback <bound method PeriodicCallback._cleanup of PeriodicCallback(callback=functools.partial(<function _reload_on_update at 0x00000250A9488AF8>, {'race_condition.py': 1632167947.1386704}), count=None, name='PeriodicCallback00196', period=500, running=False, timeout=None)> failed with following error: 'DocumentCallbackManager' object has no attribute '_change_callbacks'
```

**Screenshot**


**Screencast**
https://user-images.githubusercontent.com/19758978/134068656-65f0a758-39b8-47d6-8952-3e358cf988e7.mp4

**Pull request merged**